src/server/helpers/wikimedia.ts

```diff
@@ -0,0 +1,166 @@
+};
+
+/**
+ * Fetches a list of Wikipedia articles in all available languages for the given Wikidata item.
+ * @param {string} wikidataId - Wikidata item ID.
+ */
+export async function getAvailableWikipediaArticles(wikidataId: string, {
+	forceCache = false
+} = {}): Promise<WikipediaArticle[]> {
+	const cacheKey = `wiki:articles:${wikidataId}`;
+	const cachedArticles = await getCachedJSON<WikipediaArticle[]>(cacheKey);
+
+	if (cachedArticles || forceCache) {
+		return cachedArticles || [];
+	}
+
+	const apiUrl = new URL('https://www.wikidata.org/w/api.php');
+	apiUrl.search = new URLSearchParams({
+		action: 'wbgetentities',
+		format: 'json',
+		ids: wikidataId,
+		props: 'sitelinks'
+	}).toString();
+
+	const response = await request.get(apiUrl.href)
+		.set('User-Agent', userAgent);
+	const result = response.body as WikidataSiteLinksResult;
+	const item = result.entities?.[wikidataId];
+
+	if (!item) {
+		throw new Error(`Failed to fetch Wikidata item ${wikidataId}`);
+	}
+
+	const articles = Object.values(item.sitelinks)
+		// only keep Wikipedia pages
+		.filter((link) => link.site.endsWith('wiki'))
+		.map((page) => <WikipediaArticle>({
+			// drop project suffix
+			language: page.site.replace(/wiki$/, ''),
+			title: page.title
+		}));
+
+	cacheJSON(cacheKey, articles, {expireTime: cacheMaxAge.articles});
+
+	return articles;
+}
+
+/**
+ * Tries to find a Wikipedia article for the given Wikidata item in the first preferred language which is available.
+ * @param {string} wikidataId - Wikidata item ID.
+ * @param {string[]} preferredLanguages - List of language codes, preference in descending order.
+ */
+export async function selectWikipediaPage(wikidataId: string, {
+	forceCache = false,
+	preferredLanguages = ['en']
+} = {}) {
+	const articles = await getAvailableWikipediaArticles(wikidataId, {forceCache});
+
+	let result: WikipediaArticle;
+	for (const language of uniq(preferredLanguages)) {
+		result = articles.find((page) => page.language === toLower(language));
+		if (result) {
+			break;
+		}
+	}
+
+	return result;
+}
+
+/**
+ * Fetches the page extract of the given Wikipedia article.
+ * @param {object} article - Title and language of the article.
+ */
+export async function getWikipediaExtract(article: WikipediaArticle, {
+	forceCache = false
+} = {}): Promise<WikipediaPageExtract> {
+	const cacheKey = `wiki:extract:${article.language}:${article.title}`;
+	const cachedExtract = await getCachedJSON<WikipediaPageExtract>(cacheKey);
+
+	if (cachedExtract || forceCache) {
+		return cachedExtract;
+	}
+
+	const apiUrl = new URL(`https://${article.language}.wikipedia.org/w/api.php`);
+	apiUrl.search = new URLSearchParams({
+		action: 'query',
+		format: 'json',
+		formatversion: '2',
+		prop: 'extracts',
+		// eslint-disable-next-line sort-keys -- `exintro` only allowed with `prop: 'extracts'`
+		exintro: '1',
+		redirects: '1',
+		titles: article.title
+	}).toString();
+
+	const response = await request.get(apiUrl.href)
```

Comment:
Same here, the request can fail and take the site down with it (arguably in edge-cases, but still…)

Reply:
Also handled by the same catch blocks at router level as per my previous comment.